# Assignment 2

Hypothesis Testing

Submitted By - Subrat Kumar Nanda

Course - M.Tech 2 Yrs CSE WSPL in Data Science

Current Organization - Oracle India Pvt Lmt

In [ ]:
%matplotlib inline
import pandas as pd
import numpy as np
import statistics
import seaborn as sns
import matplotlib.pyplot as plt
from scipy import stats


In [ ]:
# Loading Data
df = pd.read_excel(r"/content/sample_data/BST_Assignment_Dataset.xlsx", sheet_name = 'Problem 1 Dataset')
df.head()


House number  Lender assessment  Borrower assessment
0             1               24.3                 18.6
1             2               31.1                 21.8
2             3              108.5                 98.1
3             4               20.0                 10.2
4             5               58.2                 50.2

In [ ]:
LA = df['Lender assessment'].to_numpy()
BA = df['Borrower assessment'].to_numpy()

In [ ]:
LA

array([ 24.3,  31.1, 108.5,  20. ,  58.2,  23.6,  38.7,  54.2,  21.3,
       145.3, 123.4, 171. ,  41.2, 123.1,  47.4,  26.1,  76.9,  52.5,
       101.2,  33.6])

In [ ]:
BA

array([ 18.6,  21.8,  98.1,  10.2,  50.2,  15.7,  29.8,  45.5,  14.6,
       135.8, 111.4, 156.5,  31.2, 109.7,  39.7,  18.6,  67.5,  42.2,
        90. ,  26.4])

In [ ]:
print(np.var(LA), np.var(BA))

2031.5886000000003 1882.659875


In [ ]:
if (np.var(LA) > np.var(BA)):
  print(np.var(LA) / np.var(BA))
elif (np.var(BA) > np.var(LA)):
  print(np.var(BA) / np.var(LA))
else:
  print("variances are equal")

1.0791054863268916


The ratio of the larger sample variance to the smaller sample variance  is less than 4. 

This means we can assume that the population variances are **equal**

Install Package pingouin

In [ ]:
# pip install pingouin

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 185 kB 4.8 MB/s 
     |████████████████████████████████| 9.9 MB 47.4 MB/s 
  Created wheel for pingouin: filename=pingouin-0.5.2-py3-none-any.whl size=196208 sha256=f484df4c6c6674d0377d8a6d47b7b09d039425c3aa215c149aa483a25b1aa13f
  Stored in directory: /root/.cache/pip/wheels/75/b0/89/f6866fc5accea9bec04f4c9f2a242520d6e8e7b8fcf89e139c
  Created wheel for littleutils: filename=littleutils-0.2.2-py3-none-any.whl size=7047 sha256=2e46a30bfdaf2b887be5e131df6948b66adafa3737209ab76cbcb8816955092e
  Stored in directory: /root/.cache/pip/wheels/6a/33/c4/0ef84d7f5568c2823e3d63a6e08988852fb9e4bc822034870a
Successfully built pingouin littleutils
  Attempting uninstall: statsmodels
    Found existing installation: statsmodels 0.12.2
    Uninstalling statsmodels-0.12.2:
      Successfully uninstalled statsmodels-0.12.2


In [ ]:

# Python program to conduct two-sample
# T-test using pingouin library
 
# Importing library
import scipy
import scipy.stats

from statsmodels.stats.weightstats import ttest_ind
import pingouin as pg

In [ ]:
#@title
#perform two sample t-test with equal variances
stats.ttest_ind(LA, BA, equal_var=True)

Ttest_indResult(statistic=0.6552566012315009, pvalue=0.5162499779174916)

### Question 1 - [a-d]
a) Obtain a 95% confidence interval for the difference of the lender’s assessment from the borrower’s assessment. What can you infer?

b) State the null and the alternative hypotheses to test whether the lender’s 
assessment is higher than the borrower’s assessment.

c) Plot the data and comment if the use of a t-test can be justified.

d) Calculate the value of appropriate test statistic and calculate the 
corresponding p-value. You may use the R function t.test. What is your 
conclusion at the 5% level of significance?

Problem Statement : Homes are typically appraised before sale. Appraisers 
hired by lenders such as banks have an incentive to assign a higher value to a 
house (so the home loan will be larger), while borrowers’ appraisers might be 
inclined to value the same house at a lower price. The article “Distressed 
Properties: Valuation Bias and Accuracy” (J. Real Estate Fin. Econ. 2010) 
describes a study in which a random sample of 20 residential properties being 
purchased in New Orleans after foreclosure was selected. Each property was 
appraised both by the borrower and by the lender, resulting in the data (in 
thousands of dollars) given in the accompanying Excel file.


In [ ]:
# Conducting two-sample ttest
result = pg.ttest(LA,BA,correction=True)

# Print the result
result

T        dof alternative     p-val            CI95%  cohen-d  \
T-test  0.655257  37.945069   two-sided  0.516256  [-19.65, 38.46]  0.20721   

         BF10     power  
T-test  0.366  0.097906

**p-value<α:**

  ⇒ reject null hypothesis (H0) 

  ⇒ accept alternative hypothesis: (HA)


​

**p-value≥α:**

⇒fail to reject null hypothesis (H0) 
​
 
​
 

Interpreting the result

This is the time to analyze the result. The p-value of the test comes out to be equal to **0.516256**, which is greater than the significance level alpha (that is, 0.05). 

P > 0.05 is the probability that the null hypothesis is true.

Also, the Cohen’s D that is obtained in a t-test is in terms of the relative strength. According to Cohen:

cohen-d = 0.2 is considered as the ‘small’ effect size

cohen-d = 0.5 is considered as the ‘medium’ effect size

cohen-d = 0.8 is considered as the ‘large’ effect size

It implies that even if the two data groups’ means don’t differ by 0.2 standard deviations or more then the difference is trivial, even if it is statistically significant.

### Question 1 - e [Wilcoxon signed-rank test]

e) Stating the null and alternative hypotheses clearly, perform a Wilcoxon signed-rank test at the 5% level of significance in the context of this problem.

In [ ]:
stats.wilcoxon(LA, BA)

WilcoxonResult(statistic=0.0, pvalue=1.9073486328125e-06)

In [ ]:
stat, pvalue = stats.wilcoxon(LA, BA)
alpha = 0.05
if pvalue > alpha:
 print('Same distribution (fail to reject H0)')
else:
 print('Different distribution (reject H0)')

Different distribution (reject H0)


Interpretation:

When P-value ≤ α :

*     Reject H0 (null hypothesis) 
*     The two groups are not identically distributed.


When P-value > α :
*     Accept H0 (null hypothesis)
*     The two groups are identically distributed.




Here, the p-value is 1.9073486328125e-06 which is less than the threshold(0.05) which is our significance level α = 0.05 i.e. p-value < alpha, reject the null hypothesis.


### Question-02

Problem Statement : In order to compare the cleansing action of three brands 
of detergents, 24 swatches of white cloth were soiled with red wine and grass 
stains and then washed with the respective detergents. The whiteness readings 
obtained are given in the accompanying Excel file. At 5% level of significance, 
test whether there are some differences in the average whitening effectiveness 
of the three detergents. Write the hypotheses, and the steps involved clearly

One-Way ANOVA in Python: One-way ANOVA (also known as “analysis of variance”) is a test that is used to find out whether there exists a statistically significant difference between the mean values of more than one group.

A one-way ANOVA has the below given null and alternative hypotheses:

H0 (null hypothesis): μ1 = μ2 = μ3  (It implies that the means of all the population are equal)

H1 (null hypothesis): It states that there will be at least one population mean that differs from the rest


Install Package researchpy

In [ ]:
# pip install researchpy

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [ ]:
import researchpy as rp

In [ ]:
# Loading Data
df = pd.read_excel(r"/content/sample_data/BST_Assignment_Dataset.xlsx", sheet_name = 'Problem 2 Dataset')
df.head()


Whiteness Reading    Detergent
0                 84  Detergent 1
1                 79  Detergent 1
2                 87  Detergent 1
3                 85  Detergent 1
4                 94  Detergent 1

In [ ]:
d1 = df.loc[df['Detergent'] == 'Detergent 1', 'Whiteness Reading'].to_numpy()
d2 = df.loc[df['Detergent'] == 'Detergent 2', 'Whiteness Reading'].to_numpy()
d3 = df.query('Detergent=="Detergent 3"')['Whiteness Reading'].to_numpy()

In [ ]:
print(d1)
print(d2)
print(d3)

[84 79 87 85 94 89 89 83]
[78 74 81 86 86 89 69 79]
[87 80 91 77 78 79 77 78]


In [ ]:
rp.summary_cont(df['Whiteness Reading'])

Variable     N     Mean      SD      SE  95% Conf.  Interval
0  Whiteness Reading  24.0  82.4583  5.9926  1.2232    79.9279   84.9888

In [ ]:
rp.summary_cont(df['Whiteness Reading'].groupby(df['Detergent']))

N    Mean      SD      SE  95% Conf.  Interval
Detergent                                                  
Detergent 1  8  86.250  4.5591  1.6119    82.4385   90.0615
Detergent 2  8  80.250  6.7135  2.3736    74.6374   85.8626
Detergent 3  8  80.875  5.2219  1.8462    76.5094   85.2406

In [ ]:
# ONE-WAY ANOVA USING SCIPY.STATS

stats.f_oneway(df['Whiteness Reading'][df['Detergent'] == 'Detergent 1'],
               df['Whiteness Reading'][df['Detergent'] == 'Detergent 2'],
               df['Whiteness Reading'][df['Detergent'] == 'Detergent 3'])

F_onewayResult(statistic=2.8040268456375834, pvalue=0.08330382794315012)

In [ ]:
stat, pvalue = stats.f_oneway(df['Whiteness Reading'][df['Detergent'] == 'Detergent 1'],
               df['Whiteness Reading'][df['Detergent'] == 'Detergent 2'],
               df['Whiteness Reading'][df['Detergent'] == 'Detergent 3'])

if pvalue > alpha:
 print('Same distribution (fail to reject H0)')
else:
 print('Different distribution (reject H0)')

Same distribution (fail to reject H0)


Interpretation

The F statistic and p-value turn out to be equal to **2.804** and **0.0833** respectively. 

Since the p-value is grater than 0.05 hence we would accept the null hypothesis. 